arch/arm64/src/imx9/imx9_usdhc.c

```diff
@@ -0,0 +1,3364 @@
+{
+  struct imx9_dev_s *priv = (struct imx9_dev_s *)dev;
+
+  mcinfo("eventset: %02x\n", eventset);
+  DEBUGASSERT(priv != NULL);
+
+  priv->cbevents = eventset;
+  imx9_callback(priv);
+}
+
+/****************************************************************************
+ * Name: imx9_registercallback
+ *
+ * Description:
+ *   Register a callback that that will be invoked on any media status
+ *   change.  Callbacks should not be made from interrupt handlers, rather
+ *   interrupt level events should be handled by calling back on the work
+ *   thread.
+ *
+ *   When this method is called, all callbacks should be disabled until they
+ *   are enabled via a call to SDIO_CALLBACKENABLE
+ *
+ * Input Parameters:
+ *   dev      - Device-specific state data
+ *   callback - The function to call on the media change
+ *   arg      - A caller provided value to return with the callback
+ *
+ * Returned Value:
+ *   0 on success; negated errno on failure.
+ *
+ ****************************************************************************/
+
+static int imx9_registercallback(struct sdio_dev_s *dev,
+                                  worker_t callback, void *arg)
+{
+  struct imx9_dev_s *priv = (struct imx9_dev_s *)dev;
+
+  /* Disable callbacks and register this callback and is argument */
+
+  mcinfo("Register %p(%p)\n", callback, arg);
+
+  DEBUGASSERT(priv != NULL);
+  priv->cbevents = 0;
+  priv->cbarg    = arg;
+  priv->callback = callback;
+  return OK;
+}
+
+/****************************************************************************
+ * Name: imx9_dmarecvsetup
+ *
+ * Description:
+ *   Setup to perform a read DMA.  If the processor supports a data cache,
+ *   then this method will also make sure that the contents of the DMA memory
+ *   and the data cache are coherent.  For read transfers this may mean
+ *   invalidating the data cache.
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *   buffer - The memory to DMA from
+ *   buflen - The size of the DMA transfer in bytes
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_IMX9_USDHC_DMA
+static int imx9_dmarecvsetup(struct sdio_dev_s *dev,
+                              uint8_t *buffer, size_t buflen)
+{
+  struct imx9_dev_s *priv = (struct imx9_dev_s *)dev;
+  DEBUGASSERT(priv != NULL && buffer != NULL && buflen > 0);
+  DEBUGASSERT(((uint64_t) buffer & 3) == 0);
+
+  /* Begin sampling register values */
+
+  imx9_sampleinit();
+  imx9_sample(priv, SAMPLENDX_BEFORE_SETUP);
+
+#if !defined(CONFIG_ARM64_DCACHE_DISABLE)
+  if (((uintptr_t)buffer & (ARMV8A_DCACHE_LINESIZE - 1)) != 0 ||
+       (buflen & (ARMV8A_DCACHE_LINESIZE - 1)) != 0)
+    {
+      /* The read buffer is not cache-line aligned. Read to an internal
+       * buffer instead.
+       */
+
+      up_invalidate_dcache((uintptr_t)priv->rxbuffer,
```

Comment:
If you want to optimize, this is actually unnecessary. You can as well just invalidate the whole rxbuffer once in the init, and afterwards every time after data has been received. Since internal rxbuffer is never written to, it is always either clean or invalid.

It is fine to have this here as well, if you like though.

Reply:
sure, it doesn't break anything, just unnecessary code. at worst causes extra memory fetch from dram to cache if there is prefetch to cache, but probably not even that.